src/main/java/net/minecraftforge/registries/RegistryObject.java

```diff
@@ -21,323 +17,67 @@
 import java.util.function.Supplier;
 import java.util.stream.Stream;
 
-public final class RegistryObject<T> implements Supplier<T>
+/**
+ * @deprecated Use {@link DeferredHolder}
+ */
+@Deprecated(since = "1.20.1", forRemoval = true)
+public final class RegistryObject<T> extends DeferredHolder<T>
 {
-    @Nullable
-    private final ResourceLocation name;
-    @Nullable
-    private ResourceKey<T> key;
-    private final boolean optionalRegistry;
-    @Nullable
-    private T value;
-    @Nullable
-    private Holder<T> holder;
-
-    /**
-     * Factory for a {@link RegistryObject} that stores the value of an object from the provided forge registry once it is ready.
-     *
-     * @param name the name of the object to look up in the forge registry
-     * @param registry the forge registry
-     * @return a {@link RegistryObject} that stores the value of an object from the provided forge registry once it is ready
-     */
     public static <T, U extends T> RegistryObject<U> create(final ResourceLocation name, IForgeRegistry<T> registry)
```

Comment:
On a related note, should IForgeRegistry#getHolder be updated to return a DeferredHolder/similar itself?

Reply:
> should they be directed to IForgeRegistry#getHolder

No, probably not, given that `getHolder` defers to the vanilla registry, which is implemented as `Optional.ofNullable(this.byKey.get(pKey))`. This operation requires that the object be registered before the Holder can be accessed (which is usually not the case at the time of creation of DH/RO's)

> IForgeRegistry#getHolder be updated to return a DeferredHolder/similar itself

Since those methods are mostly bouncers to vanilla they will be gone after the upcoming refactor. There might be room to add an additional `getDeferredHolder` method to `IForgeRegistry` but it would just be a util that wraps `create`.